keps/sig-network/1880-service-subnet/README.md

```diff
@@ -0,0 +1,558 @@
+
+TBD
+
+* Integration Tests
+
+TBD
+
+* E2E tests 
+
+
+### Graduation Criteria
+
+The change must be transparent to the users, so there should not be any performance degradation
+or Services breaking changes. 
+
+### Upgrade / Downgrade Strategy
+
+The service-subnet and node-port range will keep using the flags for the initial configuration, it can
+happen that the service subnet or node-port range are modified after the initial configuration, in that
+case they will use the same value as configured by the flags.
+
+### Version Skew Strategy
+
+That apisever control-plane will keep using repair loops to sync current service IP allocated with
+the corresponding allocator, that guarantee that the skeweved apiservers work because they have the
+same source of truth.
+
+```
+<<[UNRESOLVED detect apiservers with different version ]>>
+
+Any update on the IPRange allocated object will not be allowed if there are apiservers without the
+same version
+QUESTION Is this even possible?
+
+<<[/UNRESOLVED]>>
+```
+
+## Implementation History
+
+### Phase I
+
+The first phase consists in introduce the new API objects as a dropin replacement of current bitmap allocators.
+This can be done just [implementing the `Allocator.Interface`](https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/allocator.go) using the new API objects:
+
+```go
+// Interface manages the allocation of IP addresses out of a range. Interface
+// should be threadsafe.
+type Interface interface {
+	Allocate(net.IP) error
+	AllocateNext() (net.IP, error)
+	Release(net.IP) error
+	ForEach(func(net.IP))
+	CIDR() net.IPNet
+
+	// For testing
+	Has(ip net.IP) bool
+}
+```
+
+There is actually a working prototype with this approach in https://github.com/kubernetes/kubernetes/pull/98280
+
+### Phase 2
+
+The next phase consists in simplifying the Serice REST API leveraging the new API changes introduced in Phase I.
+
+This buils on top of Tim Hockin changes:
+
+https://github.com/kubernetes/kubernetes/pull/96684
+https://github.com/kubernetes/kubernetes/pull/95666
+
+
+With the new hooks and the new API fields, we can use API machinery tools and primitives to deal
+with the resource allocation and simplify the Services REST Strategy.
+
+
+```
+<<[UNRESOLVED allocated resources deletion ]>>
+
+Allocated resources can be owned by the services, so the Garbage Collector guarantees the release of the
+allocated resources with the 
+
+<<[/UNRESOLVED]>>
+```
+
+## Drawbacks
+
+Compared to a bitmap this design takes more space per reservation,
+which is superior only at some level of sparseness
+
+## Alternatives
+
+The inital goal was to remove the service subnet restriction for IPv6, where only /112 subnets
+or fewer weren't allowed due to the bitma restrictions.
+However, since the main problem was the allocation model for Services, the scope has been redefined
+to change the allocation model and solve the real problem instead one of its consequences:
+
```

Comment:
I'd expect to see the API server doing little work and the kube-controller-manager deciding on details like ranges and allocations. However, splitting the responsibility mean more effort to implement so I think the kube-controller-manager option should be listed as a ruled-out option.

Does that seem reasonable?

Reply:
ahh, sorry,  new plans https://github.com/kubernetes/enhancements/pull/1881#issuecomment-789151894